Tests/GRPCCoreTests/Transport/InProcessClientTransportTests.swift

```diff
@@ -0,0 +1,330 @@
+      maximumAttempts: 10,
+      initialBackoff: .seconds(1),
+      maximumBackoff: .seconds(1),
+      backoffMultiplier: 1.0,
+      retryableStatusCodes: [.unavailable]
+    )
+    let client = InProcessClientTransport(
+      server: .init(),
+      executionConfigurations: .init(defaultConfiguration: .init(retryPolicy: retryPolicy))
+    )
+
+    try await withThrowingTaskGroup(of: Void.self) { group in
+      group.addTask {
+        try await client.withStream(descriptor: .init(service: "test", method: "test")) { _ in
+          // Once the pending stream is opened, close the client to new connections,
+          // so that, once this closure is executed and this stream is closed,
+          // the client will return from `connect(lazily:)`.
+          client.close()
+        }
+      }
+
+      group.addTask {
+        // Add a sleep to make sure connection happens after `withStream` has been called,
+        // to test pending streams are handled correctly.
+        try await Task.sleep(for: .milliseconds(100))
+        try await client.connect(lazily: false)
+      }
+
+      try await group.waitForAll()
+    }
+  }
+
+  func testOpenStreamWhenClosed() async {
+    let retryPolicy = RetryPolicy(
+      maximumAttempts: 10,
+      initialBackoff: .seconds(1),
+      maximumBackoff: .seconds(1),
+      backoffMultiplier: 1.0,
+      retryableStatusCodes: [.unavailable]
+    )
+    let client = InProcessClientTransport(
+      server: .init(),
+      executionConfigurations: .init(defaultConfiguration: .init(retryPolicy: retryPolicy))
+    )
+
+    client.close()
+
+    await XCTAssertThrowsRPCErrorAsync {
+      try await client.withStream(descriptor: .init(service: "test", method: "test")) { _ in }
+    } errorHandler: { error in
+      XCTAssertEqual(error.code, .failedPrecondition)
+    }
+  }
+
+  func testOpenStreamSuccessfullyAndThenClose() async throws {
+    let retryPolicy = RetryPolicy(
+      maximumAttempts: 10,
+      initialBackoff: .seconds(1),
+      maximumBackoff: .seconds(1),
+      backoffMultiplier: 1.0,
+      retryableStatusCodes: [.unavailable]
+    )
+    let server = InProcessServerTransport()
+    let client = InProcessClientTransport(
+      server: server,
+      executionConfigurations: .init(defaultConfiguration: .init(retryPolicy: retryPolicy))
+    )
+
+    let receivedMessages = LockedValueBox([[UInt8]]())
+
+    try await withThrowingTaskGroup(of: Void.self) { group in
+      group.addTask {
+        try await client.connect(lazily: false)
+      }
+
+      group.addTask {
+        try await client.withStream(descriptor: .init(service: "test", method: "test")) { stream in
+          try await stream.outbound.write(.message([1]))
+          for try await response in stream.inbound {
+            guard case .message(let message) = response else {
+              XCTFail()
+              fatalError()
+            }
+            receivedMessages.withLockedValue({ $0.append(message) })
```

Comment:
I think this over complicates things a bit, we can just collect all inbound messages (`stream.inbound.collect()`) and check we receive the messages we expected. You'll need to close the outbound stream too.